experimental/web_client/src/grpc_web.rs

```diff
@@ -0,0 +1,104 @@
+//
+// Copyright 2022 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+//! A browser compatible gRPC web client for unary requests.
+
+use anyhow::{anyhow, Context};
+use bytes::{Buf, BufMut, Bytes, BytesMut};
+
+/// Invoke unary gRPC-web service.
+pub async fn grpc_web_unary<A: prost::Message, B: Default + prost::Message>(
+    uri: &str,
+    message: A,
+) -> anyhow::Result<B> {
+    let request_bytes = encode_body(message).context("failed to encode message")?;
+    let response_bytes = send(uri, request_bytes)
+        .await
+        .context("failed to send message")?;
+    let reply = decode_body::<B>(response_bytes)
+        .await
+        .context("failed to decode message")?;
+    Ok(reply)
+}
+
+async fn send(uri: &str, request_bytes: bytes::Bytes) -> anyhow::Result<bytes::Bytes> {
+    let client = reqwest::Client::new();
+    let response = client
+        .post(uri)
+        .header(reqwest::header::CONTENT_TYPE, "application/grpc-web")
+        .header("x-grpc-web", "1")
+        .body(request_bytes)
+        .send()
+        .await
+        .map_err(|error| anyhow!("Couldn't get response {:?}", error))?
+        .bytes()
+        .await
+        .map_err(|error| anyhow!("Couldn't get response bytes {:?}", error))?;
+    Ok(response)
+}
+
+// One byte for the compression flag plus four bytes for the length.
+// Ref: https://github.com/grpc/grpc/blob/8558f46d35cedc3ea31787aebf8d9cb07a3fc547/doc/PROTOCOL-HTTP2.md
+const GRPC_HEADER_SIZE: usize = 5;
+
+// Based off https://github.com/hyperium/tonic/blob/91b73f9fc3c1bc281e85177808721b3efe37ece0/examples/src/grpc-web/client.rs
+fn encode_body<T>(msg: T) -> anyhow::Result<Bytes>
+where
+    T: prost::Message,
+{
+    let mut buf = BytesMut::with_capacity(1024);
+
+    // first skip past the header
+    // cannot write it yet since we don't know the size of the
+    // encoded message
+    buf.reserve(GRPC_HEADER_SIZE);
+    unsafe {
+        buf.advance_mut(GRPC_HEADER_SIZE);
+    }
```

Comment:
Not sure why we need unsafe code here, can we not just write some placeholder of that size (e.g. `[0; GRPC_HEADER_SIZE]`), and then go back to it later?

Reply:
We probably could yes, I think it was done that way just to avoid the extra write operations. 
That bit comes pretty much straight from https://github.com/hyperium/tonic/blob/91b73f9fc3c1bc281e85177808721b3efe37ece0/examples/src/grpc-web/client.rs. 

Do you think it's worth changing it?

Reply:
Yes I think it's worth removing the unsafe, otherwise we would always stop and think why we have a bit of unsafe code here (which we should -- we don't want to casually skip over unsafe code while reading, it should be reserved for things that cannot be done any other way)